distributed/scheduler.py

```diff
@@ -4648,25 +4659,154 @@ async def add_nanny(self, comm):
         actors=None,
         fifo_timeout=0,
         code=None,
+        workers=None,
+        allow_other_workers=None,
+        annotations=None,
     ):
-        unpacked_graph = HighLevelGraph.__dask_distributed_unpack__(hlg)
-        dsk = unpacked_graph["dsk"]
-        dependencies = unpacked_graph["deps"]
-        annotations = unpacked_graph["annotations"]
+        tok = _current_client.set(False)  # type: ignore
+        try:
+
+            graph: HighLevelGraph = pickle.loads(graph_header, buffers=graph_frames)
+        except Exception as e:
+            _current_client.reset(tok)
+            text = str(e)
+            exc = pickle.dumps(e)
+            parent: SchedulerState = cast(SchedulerState, self)
+            for key in keys:
+                ts = parent.new_task(
+                    key, None, "erred", computation=None  # computation
+                )
+                ts._exception = exc
+                ts._exception_text = text
+                ts._exception_blame = ts
+                parent.tasks[key] = ts
+
+            self.client_desires_keys(keys=keys, client=client)
+            return
+
+        if annotations is None:
+            annotations = {}
+
+        workers = workers or annotations.pop("workers", None)
+        allow_other_workers = allow_other_workers or annotations.pop(
+            "allow_other_workers", None
+        )
+        if retries is None:
+            retries = annotations.pop("retries", None)
+        resources = resources or annotations.pop("resources", None)
+        if user_priority is None:
+            user_priority = annotations.pop("priority", None)
+
+        if isinstance(workers, (str, Number)):
+            workers = [workers]
+        if isinstance(workers, (tuple, set)):
+            workers = list(workers)
+        if isinstance(workers, list):
+            restrictions = workers
+        elif workers is None:
+            restrictions = []
+        else:
+            raise TypeError("Workers must be a list or set of workers or None")
+
+        dsk = dict(graph)
+
+        if allow_other_workers:
+            loose_restrictions = set(dsk)
+        else:
+            loose_restrictions = set()
+
+        from distributed.utils_comm import unpack_remotedata
+
+        dependencies, dependents = get_deps(dsk)
+
+        # Remove `Future` objects from graph and note any future     dependencies
+        dsk2 = {}
+        fut_deps = {}
+        for k, v in dsk.items():
+            dsk2[k], futs = unpack_remotedata(v, byte_keys=True)
+            if futs:
+                fut_deps[k] = futs
+        dsk = dsk2
+
+        # - Add in deps for any tasks that depend on futures
+        for k, futures in fut_deps.items():
+            dependencies[k].update(f.key for f in futures)
+
+        pre_stringify = set(dsk)
+        dsk = {stringify(k): stringify(v, exclusive=graph) for k, v in dsk.items()}
+
+        def process(x, keys=dsk, string_keys=pre_stringify):
```

Comment:
It sure would be nice if we didn't need to worry about stringifying keys anymore. Is the historical reason for doing this related to lists and tuples being indistinguishable in msgpack, or are there other reasons?

Reply:
There may be a performance reason, but in general I agree that we might want to remove stringification.  I would propose to handle that in follow-up work though if we do that.

Reply:
I see. If stringification isn't an actual barrier here, then I completely agree with addressing that separately.